File: leetcode/rotate_array.py
from typing import List


class Solution:
    def rotate(self, nums: List[int], k: int) -> List:
        for _ in range(k):
            self.reverse(nums)

        return nums
    
    def reverse(self, nums: List[int]) -> List:
        i, end = 0, len(nums) - 1
        while i < end:
            nums[i], nums[end] = nums[end], nums[i]
            i += 1

        return nums
    
if __name__ == "__main__":
    s = Solution()
    nums = [1, 2, 3, 4, 5]
    k = 3
    print(s.rotate(nums, k))

File: leetcode/rotate_array.cpp
using namespace std;
#include <iostream>
#include <vector>


class Solution {
public: 
    void reverse(vector<int>::iterator start, vector<int>::iterator end) {
        while (start < end) {
            int temp = *start;
            *start = *end;
            *end = temp;
            start++;
            end--;
        }
    }

    void rotate(vector<int>& nums, int k) {
        int n = nums.size();
        k = k % n;

        reverse(nums.begin(), nums.end()); // reverse the whole array
        reverse(nums.begin(), nums.begin() + k); // reverse the first k elements
        reverse(nums.begin() + k, nums.end()); // reverse the last k elements
    }
};

int main() {
    Solution so;
    vector<int> nums = {1, 2, 3, 4, 5};
    int k = 3;

    so.rotate(nums, k);

    for (int num : nums) {
        cout << num << " ";
    }
    cout << endl;
    
    return 0;
}